zerver/lib/actions.py

```diff
@@ -1923,23 +1925,24 @@ def get_recipient_from_user_ids(recipient_profile_ids: Set[int],
         # for whether forwarder_user_profile is among the private
         # message recipients of the message.
         assert forwarder_user_profile is not None
-        if forwarder_user_profile.id not in recipient_profile_ids:
+        if forwarder_user_profile not in recipient_profiles:
             raise ValidationError(_("User not authorized for this query"))
 
-    if (len(recipient_profile_ids) == 2 and sender.id in recipient_profile_ids):
-        recipient_profile_ids.remove(sender.id)
+    if (len(recipient_profiles) == 2 and sender in recipient_profiles):
+        recipient_profiles.remove(sender)
 
-    if len(recipient_profile_ids) > 1:
+    if len(recipient_profiles) > 1:
         # Make sure the sender is included in huddle messages
-        recipient_profile_ids.add(sender.id)
-        return get_huddle_recipient(recipient_profile_ids)
+        recipient_profiles.add(sender)
```

Comment:
So, as a policy matter we've been avoiding adding UserProfile objects (or other Django model objects) to sets, because while it can appear to work, the behavior isn't super well-defined.  We should probably replace the `set` data structure with a `dict[int, UserProfile]` mapping or something similar.